### PR TITLE
fix the ptr.annotation generation in reverse translator

### PR DESCRIPTION
**Pull request by broxigarchen**
Fix the `ptr.annotation` in reverse translator such that the generated pointer annotation call is being used properly in the IR.

The IR generated from the previous code base is:
```
%0 = xxx
%1 = ptr.annotation %0 ...  (left unused)
%2 = load %0 ...
```
while the expected form is:
```
%0 = xxx
%1 = ptr.annotation %0 ...
%2 = load %1 ...
```

### CODE REVIEW

**Review by MrSidims — COMMENTED** [review]
I though that there was/is a patch from @vmaksimo to fix annotations on dead code path

**Comment by broxigarchen** [comment]
> I though that there was/is a patch from @vmaksimo to fix annotations on dead code path

Hi @vmaksimo can you comment? I am posting this PR since the annotated_ptr feature that I am working on is pending on it

**Comment by vmaksimo** [comment]
Hi @broxigarchen. I did the change with the similar impact in January https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1723
But I guess your patch could be more general, I somehow tested my PR only for clang::annotate (UserSemantic decorations).

Could you please add/modify some test so we can see which case is not covered by the current implementation?

**Comment by broxigarchen** [comment]
> Hi @broxigarchen. I did the change with the similar impact in January #1723 But I guess your patch could be more general, I somehow tested my PR only for clang::annotate (UserSemantic decorations).
> 
> Could you please add/modify some test so we can see which case is not covered by the current implementation?

Sure. Let me update the lit test that was failing without this patch

**Comment by broxigarchen** [comment]
> Hi @broxigarchen. I did the change with the similar impact in January #1723 But I guess your patch could be more general, I somehow tested my PR only for clang::annotate (UserSemantic decorations).
> 
> Could you please add/modify some test so we can see which case is not covered by the current implementation?

Done. In the updated lit test, there are multiple duplicated bit-cast auto-generated from the reverse translator, and I guess that could be why it failed to recognize the ptr.annotation before

File: lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -1236,20 +1236,67 @@ Value *SPIRVToLLVM::oclTransConstantPipeStorage(
...
+    while (Inst && (isa<BitCastInst>(Inst) || isa<AddrSpaceCastInst>(Inst))) {
+      if ((Inst = dyn_cast<Instruction>(Inst->getOperand(0))) &&
+          SearchPtrAnn(Inst, CallResult)) {
+        BaseValue = Inst;
+        break;
+      }
     }
   }
+
+  // overwite operand with intrinsic call result
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVReader.cpp:1283]:
```suggestion
  // overwrite operand with intrinsic call result
```

File: lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -1236,20 +1236,67 @@ Value *SPIRVToLLVM::oclTransConstantPipeStorage(
...
+  if (CallResult) {
+    SPIRVDBG(dbgs() << "BaseValue: " << *BaseValue << '\n'
+                    << "CallResult: " << *CallResult << '\n');
+    DominatorTree DT(*F);
+    BaseValue->replaceUsesWithIf(CallResult, [&DT, &CallResult](Use &U) {
+      return DT.dominates(CallResult, U);
+    });
+
+    // overwite V
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVReader.cpp:1292]:
```suggestion
    // overwrite V
```

**Review by MrSidims — APPROVED** [review]

**Comment by MrSidims** [comment]
Restarting CI with new head

**Pull request merged**